File: templates/arrays/sliding_window.go
package arrays

func MinSubarrayLen(target int, nums []int) int {
	left := 0
	sum := 0
	best := len(nums) + 1

	for right, v := range nums {
		sum += v
		for sum >= target {
			length := right - left + 1
			if length < best {
				best = length
			}
			sum -= nums[left]
			left++
		}
	}

	if best == len(nums)+1 {
		return 0
	}
	return best
}

func LongestSubstringWithoutRepeat(s string) int {
	last := map[byte]int{}
	left := 0
	best := 0

	for right := 0; right < len(s); right++ {
		if idx, ok := last[s[right]]; ok && idx >= left {
			left = idx + 1
		}
		last[s[right]] = right
		if right-left+1 > best {
			best = right - left + 1
		}
	}

	return best
}

func MaxSumFixedWindow(nums []int, k int) int {
	if k <= 0 || k > len(nums) {
		return 0
	}

	sum := 0
	for i := 0; i < k; i++ {
		sum += nums[i]
	}
	best := sum

	for i := k; i < len(nums); i++ {
		sum += nums[i] - nums[i-k]
		if sum > best {
			best = sum
		}
	}
	return best
}

File: templates/arrays/sliding_window.py
def min_subarray_len(target: int, nums: list[int]) -> int:
    left = 0
    total = 0
    best = len(nums) + 1

    for right, value in enumerate(nums):
        total += value
        while total >= target:
            best = min(best, right - left + 1)
            total -= nums[left]
            left += 1

    return 0 if best == len(nums) + 1 else best


def longest_substring_without_repeat(text: str) -> int:
    last: dict[str, int] = {}
    left = 0
    best = 0

    for right, ch in enumerate(text):
        if ch in last and last[ch] >= left:
            left = last[ch] + 1
        last[ch] = right
        best = max(best, right - left + 1)

    return best


def max_sum_fixed_window(nums: list[int], k: int) -> int:
    if k <= 0 or k > len(nums):
        return 0

    total = sum(nums[:k])
    best = total

    for i in range(k, len(nums)):
        total += nums[i] - nums[i - k]
        best = max(best, total)

    return best
